core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,247 @@
+import org.apache.spark.Success
+import org.apache.spark.TaskEndReason
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.util.SystemClock
+import org.apache.spark.util.ThreadUtils
+import org.apache.spark.util.Utils
+import org.apache.spark.util.Clock
+
+import com.google.common.annotations.VisibleForTesting
+
+/**
+ * BlacklistTracker is design to track problematic executors and node on application level.
+ * It is shared by all TaskSet, so that once a new TaskSet coming, it could be benefit from
+ * previous experience of other TaskSet.
+ *
+ * Once task finished, the callback method in TaskSetManager should update
+ * executorIdToFailureStatus Map.
+ */
+private[spark] class BlacklistTracker(
+    sparkConf: SparkConf,
+    clock: Clock = new SystemClock()) extends BlacklistCache{
+  // maintain a ExecutorId --> FailureStatus HashMap
+  private val executorIdToFailureStatus: mutable.HashMap[String, FailureStatus] = mutable.HashMap()
+
+  // Apply Strategy pattern here to change different blacklist detection logic
+  private val strategy = BlacklistStrategy(sparkConf)
+
+  // A daemon thread to expire blacklist executor periodically
+  private val scheduler = ThreadUtils.newDaemonSingleThreadScheduledExecutor(
+      "spark-scheduler-blacklist-expire-timer")
+
+  private val recoverPeriod = sparkConf.getTimeAsSeconds(
+    "spark.scheduler.blacklist.recoverPeriod", "60s")
+
+  def start(): Unit = {
+    val scheduleTask = new Runnable() {
+      override def run(): Unit = {
+        Utils.logUncaughtExceptions(expireExecutorsInBlackList())
+      }
+    }
+    scheduler.scheduleAtFixedRate(scheduleTask, 0L, recoverPeriod, TimeUnit.SECONDS)
+  }
+
+  def stop(): Unit = {
+    scheduler.shutdown()
+    scheduler.awaitTermination(10, TimeUnit.SECONDS)
+  }
+
+  // The actual implementation is delegated to strategy
+  /** VisibleForTesting */
+  private[scheduler] def expireExecutorsInBlackList(): Unit = synchronized {
+    val updated = strategy.expireExecutorsInBlackList(executorIdToFailureStatus, clock)
+    if (updated) {
+      invalidAllCache()
+    }
+  }
+
+  def updateFailureExecutors(
+      stageId: Int, info: TaskInfo, reason: TaskEndReason) : Unit = synchronized {
+    val index = info.index
+    val atomTask = BlacklistAtomTask(stageId, index)
+    reason match {
+      // If task succeeding, remove related record from executorIdToFailureStatus
+      case Success =>
+        removeFailureExecutorsForTaskId(info.executorId, stageId, index)
+
+      // If task failing, update latest failure time and failedTaskIds
+      case _ =>
+        val executorId = info.executorId
+        executorIdToFailureStatus.get(executorId) match {
+          case Some(failureStatus) =>
+            failureStatus.updatedTime = clock.getTimeMillis()
+            val failedTimes = failureStatus.numFailuresPerTask.getOrElse(atomTask, 0) + 1
+            failureStatus.numFailuresPerTask.update(atomTask, failedTimes)
```

Comment:
nit: rather than calling `update`, its more idiomatic to write this as `failureStatus.numFailuresPerTask(atomTask) = failedTimes` (scala syntax-sugar for the `update` method)